cpp/src/arrow/python/pandas_convert.cc

```diff
@@ -1379,7 +1375,11 @@ class DatetimeBlock : public PandasBlock {
 
     const ChunkedArray& data = *col.get()->data();
 
-    if (type == Type::DATE64) {
+    if (type == Type::DATE32) {
+      // Date64Type is millisecond timestamp stored as int64_t
+      // TODO(wesm): Do we want to make sure to zero out the milliseconds?
+      ConvertDatetimeNanos<int32_t, 86400000000000LL>(data, out_buffer);
```

Comment:
And generally speaking to have a header with some constants for readability with things like `npy_shift`.